src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -501,37 +500,38 @@ class ConcatExpression : public detail::VariadicExpression {
         size_t i = 0;
         for (auto& el : gen) {
           auto literal = valueGetter(std::move(el), ctx);
-          concatOrSetLiteral(resultAsVec[i], literal, isInitializedFlag);
+          concatOrSetLiteral(resultAsVec[i], literal, isFirstLiteral);
           ctx->cancellationHandle_->throwIfCancelled();
           ++i;
         }
       }
       ctx->cancellationHandle_->throwIfCancelled();
     };
-    ql::ranges::for_each(
-        childrenVec(), [&ctx, &visitSingleExpressionResult](const auto& child) {
-          std::visit(visitSingleExpressionResult, child->evaluate(ctx));
-        });
+    ql::ranges::for_each(childrenVec(), [&ctx, &visitSingleExpressionResult,
+                                         &isFirstLiteral](const auto& child) {
+      std::visit(visitSingleExpressionResult, child->evaluate(ctx));
+      isFirstLiteral = false;
+    });
 
     // Lift the result from `string` to `IdOrLiteralOrIri` which is needed for
     // the expression module.
 
     auto visitLiteralResult =
-        [&](const std::optional<Literal>& literalSoFar) -> ExpressionResult {
-      return convertLiteral(literalSoFar);
+        [&](std::optional<Literal>& literalSoFar) -> ExpressionResult {
+      return moveLiteralToResult(literalSoFar);
     };
 
     auto visitLiteralVecResult =
-        [&](const LiteralVec& literalVec) -> ExpressionResult {
+        [&](LiteralVec& literalVec) -> ExpressionResult {
       VectorWithMemoryLimit<IdOrLiteralOrIri> resultAsVec(ctx->_allocator);
       resultAsVec.reserve(literalVec.size());
-      ql::ranges::copy(literalVec | ql::views::transform(convertLiteral),
+      ql::ranges::copy(literalVec | ql::views::transform(moveLiteralToResult),
                        std::back_inserter(resultAsVec));
       return resultAsVec;
     };
     return std::visit(ad_utility::OverloadCallOperator{visitLiteralResult,
                                                        visitLiteralVecResult},
-                      std::move(result));
+                      result);
```

Comment:
That is fine, if you take a (non-const) `&` , then you cannot move into it, 
But your lambdas move anyway, so that is okay here as is.